File: scraper/functions/publish/storage.ts
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import type { DbSchema, Job } from '../../core/types';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const PROJECT_ROOT = path.resolve(__dirname, '..', '..', '..');
const DB_DIR = path.join(PROJECT_ROOT, 'data');
const DB_FILE = path.join(DB_DIR, 'db.json');

let cachedDb: DbSchema | null = null;

function readDb(): DbSchema {
  if (cachedDb) return cachedDb;
  try {
    const data = fs.readFileSync(DB_FILE, 'utf-8');
    cachedDb = JSON.parse(data);
  } catch {
    cachedDb = null;
  }
  return cachedDb || { users: [], jobs: [], applications: [] };
}

function writeDb(db: DbSchema): void {
  if (!fs.existsSync(DB_DIR)) {
    fs.mkdirSync(DB_DIR, { recursive: true });
  }
  fs.writeFileSync(DB_FILE, JSON.stringify(db, null, 2), 'utf-8');
  cachedDb = db;
}

export function mergeJobs(newJobs: Job[]): { inserted: number; updated: number } {
  const db = readDb();
  let inserted = 0;
  let updated = 0;

  for (const job of newJobs) {
    const existingIndex = job.sourceId
      ? db.jobs.findIndex(j => j.sourceId === job.sourceId)
      : -1;

    if (existingIndex >= 0) {
      db.jobs[existingIndex] = { ...db.jobs[existingIndex], ...job };
      updated++;
    } else {
      db.jobs.push(job);
      inserted++;
    }
  }

  writeDb(db);
  return { inserted, updated };
}

export function cleanupJobs(daysOld: number = 7): { removed: number; removedManual: number } {
  const db = readDb();
  const cutoff = new Date();
  cutoff.setDate(cutoff.getDate() - daysOld);
  let removed = 0;
  let removedManual = 0;

  const before = db.jobs.length;
  db.jobs = db.jobs.filter(j => {
    if (!j.source) { removedManual++; return false; }
    const d = new Date(j.createdAt);
    if (isNaN(d.getTime()) || d < cutoff) { removed++; return false; }
    return true;
  });

  if (db.jobs.length !== before) {
    writeDb(db);
  }
  return { removed, removedManual };
}

export function getAllJobs(): Job[] {
  return readDb().jobs;
}

export function getStats() {
  const db = readDb();
  const sourceCounts: Record<string, number> = {};
  for (const j of db.jobs) {
    sourceCounts[j.source || 'manual'] = (sourceCounts[j.source || 'manual'] || 0) + 1;
  }
  return {
    totalJobs: db.jobs.length,
    sourceCounts,
  };
}

File: scraper/storage.py
"""Persist jobs into data/db.json: merge by sourceId, cleanup old entries."""
import json
import os
from datetime import datetime, timedelta, timezone

PROJECT_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
DB_DIR = os.path.join(PROJECT_ROOT, "data")
DB_FILE = os.path.join(DB_DIR, "db.json")

def read_db():
    try:
        with open(DB_FILE, "r", encoding="utf-8") as f:
            db = json.load(f)
    except (OSError, json.JSONDecodeError):
        return {"jobs": []}
    return {"jobs": db.get("jobs") or []}


def write_db(db):
    os.makedirs(DB_DIR, exist_ok=True)
    with open(DB_FILE, "w", encoding="utf-8") as f:
        json.dump(db, f, indent=2, ensure_ascii=False)
        f.write("\n")


def _dedup_key(job):
    # keyed by (source, sourceId): numeric ids can collide across sources
    return (job.get("source"), job.get("sourceId"))


def merge_jobs(new_jobs):
    db = read_db()
    by_key = {_dedup_key(j): i for i, j in enumerate(db["jobs"]) if j.get("sourceId")}
    inserted = 0
    updated = 0

    for job in new_jobs:
        existing_index = by_key.get(_dedup_key(job), -1)
        if existing_index >= 0:
            db["jobs"][existing_index] = {**db["jobs"][existing_index], **job}
            updated += 1
        else:
            by_key[_dedup_key(job)] = len(db["jobs"])
            db["jobs"].append(job)
            inserted += 1

    write_db(db)
    return inserted, updated


def _parse_date(value):
    if not value:
        return None
    try:
        return datetime.fromisoformat(str(value).replace("Z", "+00:00"))
    except ValueError:
        return None


def cleanup_jobs(days_old=7):
    db = read_db()
    cutoff = datetime.now(timezone.utc) - timedelta(days=days_old)
    removed = 0
    removed_manual = 0
    kept = []

    for job in db["jobs"]:
        if not job.get("source"):
            removed_manual += 1
            continue
        created = _parse_date(job.get("createdAt"))
        if created is None or created < cutoff:
            removed += 1
            continue
        kept.append(job)

    if len(kept) != len(db["jobs"]):
        db["jobs"] = kept
        write_db(db)

    return removed, removed_manual


def get_stats():
    db = read_db()
    source_counts = {}
    for job in db["jobs"]:
        source = job.get("source") or "manual"
        source_counts[source] = source_counts.get(source, 0) + 1
    return {"totalJobs": len(db["jobs"]), "sourceCounts": source_counts}
